trieste/models/gpflow/optimizer.py

```diff
@@ -92,3 +94,30 @@ def compiled_closure() -> tf.Tensor:
         builder.closure_builder = closure_builder
 
     return builder.closure_builder((X, Y))
+
+
+def loss_history_callback(
```

Comment:
I was thinking of getting the whole `OptimizeResult` object, rather than just loss: https://docs.scipy.org/doc/scipy/reference/generated/scipy.optimize.OptimizeResult.html

when we call `minimize` in gpflow.scipy, it should return it, right?
then `fun` should be actual loss value, but one gets all the other stuff that is useful for evaluating the optimisation: success, status, nfev, nit ...

Reply:
We can certainly get the OptimizeResult, but I thought that didn't include any loss history (unlike the keras optimizer) just the number of evaluations? Isn't that what you wanted?

Reply:
As an aside, how were you expecting to get the result? TrainableProbabilisticModel.optimize currently doesn't return anything (though Optimizer.optimize already does). If you're using BO or AskTell then there is no obvious way to return all the results when optimizing all the models, but I guess we could place them somewhere in the model wrappers (a bit like how the keras history is available inside DeepEnsemble.model.history).

Reply:
ok, you want to say we would get here a loss with every iteration of the optimiser? rather than a single end result with `fun` in `OptimizeResult`?

Reply:
> we could place them somewhere in the model wrappers (a bit like how the keras history is already available inside DeepEnsemble.model.history)

yes, this was the solution I was thinking of - but not sure if we can attach a history object to gpflow model like we have with keras model object, if not and we store it in the wrapper, perhaps we should have a unified way for all models, store it in the wrapper object

Reply:
> ok, you want to say we would get here a loss with every iteration of the optimiser? rather than a single end result with fun in OptimizeResult?

Yes, that's what this does. Essentially equivalent to `history.history["loss"]` on the keras result.

>  perhaps we should have a unified way for all models, store it in the wrapper object

We could potentially add some property to `TrainableProbabilisticModel` but note that the `keras.callbacks.History` is the result of calling `fit` on a keras model, not of calling `minimize` on the optimizer (which returns nothing) so the two aren't directly comparable. Will have a think.

Reply:
We could also just store the most recent result of calling `Optimizer.optimize` in the optimizer itself, which would let us easily access the OptimizeResult object. Or even all the previous results (enabled via a `save_results: bool` option in the Optimizer).

Reply:
> > ok, you want to say we would get here a loss with every iteration of the optimiser? rather than a single end result with fun in OptimizeResult?

> Yes, that's what this does. Essentially equivalent to history.history["loss"] on the keras result.

I see, that would be useful, but since it can expensive perhaps leaving off by default? (depends how expensive it is...)

Reply:
> We could also just store the most recent result of calling Optimizer.optimize in the optimizer itself, which would let us easily access the OptimizeResult object. Or even all the previous results (enabled via a save_results: bool option in the Optimizer).

I think that is the best solution, all the history from each active learning step

for keras we can simply copy it from the model attribute, just so that we have a unified way of accessing the optimization history?

Reply:
Remember that for keas the history is generated by the model calling the fit method, not the optimizer's minimize method.

Reply:
we could still attach the resulting history to the optimizer attribute, we would just need to do it from the model wrapper, in the optimize method, no?